### PR TITLE
Bluetooth read data

### LINKED ISSUE #2: Add ability to read data from Bluetooth device
### Description

Add functionality to be able to read data with a specific request from a bluetooth device. 

### Suggestions

_N/A_

### Additional Context

_N/A_

### PR DESCRIPTION
**Pull request by Fulons**
To review:
- BleWinrtDll.cpp is the meat of this review
- The C# project is only for debugging and should not need review, but can be good for understanding usage quicker

To test:
 - Make sure BleWinrtDll project builds and produces a dll (this should now have the suffix of Debug in debug builds to avoid accidentally copying wrong dll)
 - Run DebugBle, all BLE devices ever connected to your pc will be listed in the console (if you have a connectionless BLE device change the device name (program.cs:23) to scan characteristics and attempt to read values)

closes #2

### CODE REVIEW

File: DebugBle/BLE.cs [hunk 1]
```diff
@@ -7,6 +7,12 @@
 
 public class BLE
 {
+    const string DLL_NAME =
+#if DEBUG
+        "BleWinrtDllDebug.dll";
+#else
+        "BleWinrtDll.dll";
+#endif
```
Comment (SamHignett) [inline DebugBle/BLE.cs:15]: image

I could just be using it wrong, but even though I have my solution configuration set to "Debug", it is attempting to copy the non-debug version of the dll. 

Is this code snippet supposed to change which library is copied/loaded, or something else?

File: DebugBle/BLE.cs
Comment (SamHignett) [inline DebugBle/BLE.cs:1]: image

After building both the debug/release version of the dll, and copying it to the `bluetooth-windows-rt\DebugBle` folder and running the application, it seems that no devices were found and the scan instantly finished. Previously the scan was  constant and found several devices of mine that should still be on. 

Does this mean the bluetooth device scanning did not work, do I need additional setup or is this simply a feature now that it won't show all devices unless you query them or something like that?

**Review by SamHignett — CHANGES_REQUESTED** [review]

**Review by d-blackburn — DISMISSED** [review]
LGTM! 👍🏼

File: DebugBle/BLE.cs
Comment (Fulons) [reply DebugBle/BLE.cs:1]: Check under Bluetooth & other devices in windows and ensure you have a paired BLE device (should not need to be connected)

File: BleWinrtDll/BleWinrtDll.cpp [hunk 2]
```diff
@@ -560,6 +561,52 @@ bool SendData(BLEData* data, bool block) {
...
+	try {
+		auto characteristic = co_await retrieveCharacteristic(id->deviceId, id->serviceUuid, id->characteristicUuid);
+		if (characteristic != nullptr) {
+			//Retrieve data from device not from windows cache
+			auto status = co_await characteristic.ReadValueAsync(BluetoothCacheMode::Uncached);
+			if (status.Status() == GattCommunicationStatus::Success) {
+				auto value = status.Value();
+				auto valueData = value.data();
+				for (uint32_t i = 0; i < value.Length(); ++i) {
+					data->buf[i] = valueData[i];
```
[the remark this answers is not in the document]
Comment (Fulons) [reply BleWinrtDll/BleWinrtDll.cpp:575]: Made sure index wont go out of range. Also added an error message to warn when this happens.

File: DebugBle/BLE.cs (hunk 1 above)
Comment (SamHignett) [reply DebugBle/BLE.cs:15]: After running it again this morning, it seems to have copied the correct file now 👍 , no clue what changed though!

image

Just to confirm in Rider that the post-build event script is correct.

File: DebugBle/BLE.cs
Comment (SamHignett) [reply DebugBle/BLE.cs:1]: Tested again with Bluetooth headphones paired and turned on, and it is now detecting them. 

It is curious how it was seemingly finding non-paired devices previously, but regardless it correctly detected a paired bluetooth device which is the goal.

**Review by SamHignett — APPROVED** [review]
LGTM! 👍

**Pull request merged**